backend/functions/src/emailing/functions.ts

```diff
@@ -1,4 +1,50 @@
 import axios from 'axios'
+
+// firebase imports
+import admin from 'firebase-admin'
+import { db } from '../config'
+const courseRef = db.collection('courses')
+
+// ==== Timestamp helper functions
+
+type timestampsType = { [key: string]: string }
+
+const timestamps: timestampsType = {
+  'Share matched results': 'shareMatchEmailTimestamp',
+  'First no match notification': 'checkInEmailTimestamp',
+  'Second no match notification': 'firstNoMatchEmailTimestamp',
+  'Request to add student to group': 'secondNoMatchEmailTimestamp',
+  'Request to add student to group (late)': 'addStudentEmailTimestamp',
+  'Ask to join group': 'lateAddStudentEmailTimestamp',
+  'Check in with groups': 'askJoinGroupEmailTimestamp',
+}
+
+const getTimestampField = (template: string) => {
+  return timestamps[template]
+}
+
+/** Updating Email Sent Timestap:
+ * @param courseId is the string courseId usually in the form of a six-digit number such as 358546.
+ * @param group is the group for the given course that emails were sent to.
+ * @param template is the name of the template of the email being sent. Matched emailTemplates.js names made by sean. Currently is the string value, may change to the variable value in future (less wordy).
+ * @result updates database to have email sent timestamp to current time.
+ *  */
+export const updateEmailTimestamp = async (
+  courseId: string,
+  group: string,
+  template: string
+) => {
+  const time = admin.firestore.FieldValue.serverTimestamp()
+  const timestampField = getTimestampField(template)
+  // must be string format -> parse here or when calling function
+  await courseRef
+    .doc(courseId)
+    .collection('groups')
+    .doc(group)
+    .update({ [timestampField]: time })
```

Comment:
Another way of looking at this function could be to think of it as a wrapper around the Firestore call to update the timestamps, in which case you could remove the async and just return the courseRef...update(). Basically taking the parameters, processing them, and returning the Promise that is the database write. Don't think it's necessary though